### PR TITLE
user-connect

### LINKED ISSUE #9: As a user, I want to be able to connect using my private key so that I can share the public key with a website
**Background**:

Some websites require users to Connect their account before being able to show bespoke content to them.
Specific functionality might only be available if the website knows which user is connected (such as showing balances).

As such, and in line with the premise of reducing cognitive complexity, entering the private key is the first task to be completed for the /connect endpoint

**Acceptance** criteria:

[ ] A user can enter their private key (shown as a password/with the option of making it visible)
[ ] Once connected, an event is emitted to the parent window (if any) sharing the public key.
[ ] The private key using a salt value and is encrypted and stored in sessionStorage

### PR DESCRIPTION
**Pull request by Ja-boop**

### CODE REVIEW

File: src/api/stellarSdk.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+import * as StellarSDK from 'stellar-sdk';
```
Comment (fsodano) [inline src/api/stellarSdk.ts]: here you can simply do `import StellarSdk from 'stellar-sdk'`

File: src/api/stellarSdk.ts [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+const server = new StellarSDK.Server('https://horizon-testnet.stellar.org');
```
Comment (fsodano) [inline src/api/stellarSdk.ts]: this should come from an env var

File: src/routes/Connect.svelte [hunk 3]
```diff
@@ -1,4 +1,14 @@
...
+    import { changeTitle } from '../services/utils';
+
+    const toggleVisibility = () => {
+        const input = <HTMLInputElement>document.querySelector('#secret-key-input');
+        if (input.type === 'password') {
+            input.type = 'text';
+        } else {
+            input.type = 'password';
+        }
+    };
```
Comment (fsodano) [inline src/routes/Connect.svelte]: I'm sure there's a way to do this by assigning dynamic types? such as `inputType={showSecret ? 'text' : 'password'}`

File: src/api/stellarSdk.ts [hunk 4]
```diff
@@ -0,0 +1,4 @@
...
+
+export { StellarSDK, server };
```
Comment (fsodano) [inline src/api/stellarSdk.ts]: Come to think of it... why is this file necessary at all?

File: src/services/connect.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import { StellarSDK } from '../api/stellarSdk';
+import type { Keypair } from 'stellar-sdk';
+
+const encryptKey = (key: string, salt: number) => {
+    const encodedData = btoa(key);
+    const saltValue = Math.floor(Math.random() * 10000000 * salt);
```
Comment (fsodano) [inline src/services/connect.ts]: the salt value must be constant, otherwise you'll never be able to decrypt the secret :)

File: src/services/connect.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import { StellarSDK } from '../api/stellarSdk';
+import type { Keypair } from 'stellar-sdk';
+
+const encryptKey = (key: string, salt: number) => {
+    const encodedData = btoa(key);
```
Comment (fsodano) [inline src/services/connect.ts]: don't use `btoa`, give https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/encrypt a look, seems to work everywhere except IE11 for which we can polyfill it later on

File: src/services/connect.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
+import { StellarSDK } from '../api/stellarSdk';
+import type { Keypair } from 'stellar-sdk';
+
+const encryptKey = (key: string, salt: number) => {
```
Comment (fsodano) [inline src/services/connect.ts]: the salt should not be passed in as a parameter.

The `encrypt` and `decrypt` functions should exist in a utils/security.ts file rather than live in the `connect` service

File: src/services/connect.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    sessionStorage.setItem('key', encodedKey);
+};
+
+const connectWithSecretKey = (key: string): Keypair | void => {
+    try {
+        const sourceKeys = StellarSDK.Keypair.fromSecret(key);
+        saveSecretKeyInSessionStorage(key);
+        return sourceKeys;
+    } catch (e) {
+        return console.error('Invalid key: ', e);
```
Comment (fsodano) [inline src/services/connect.ts]: you should re-throw the error and therefore this function should never return `void`

File: src/services/connect.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    }
+};
+
+export const getPublicKey = (key: string): string | void => {
```
Comment (fsodano) [inline src/services/connect.ts]: this function should not return void

File: src/services/connect.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    try {
+        const sourceKeys = connectWithSecretKey(key);
+
+        if (sourceKeys) {
+            const publicKey = sourceKeys.publicKey();
+            return publicKey;
+        }
+    } catch (e) {
+        return console.log('Error while getting the public key', e);
```
Comment (fsodano) [inline src/services/connect.ts]:
```suggestion
        console.error('Error while getting the public key', e);
        throw e;
```

File: src/services/utils.ts [hunk 11]
```diff
@@ -0,0 +1,14 @@
+import { getPublicKey } from './connect';
+
+export const changeTitle = () => {
+    const input = <HTMLInputElement>document.querySelector('#secret-key-input');
+    const title = <HTMLInputElement>document.querySelector('#public-key-title');
```
Comment (fsodano) [inline src/services/utils.ts]: I don't think it's a very svelte-like way of doing things, you should make use of the reactivity by storing the public key, and also storing if there was an error or not, and have your component update automatically (i.e. react to those changes)

**Review by fsodano — CHANGES_REQUESTED** [review]

File: src/api/stellarSdk.ts (hunk 4 above)
Comment (Ja-boop) [reply src/api/stellarSdk.ts]: The purpose of this file is to keep the API call in one place so that the rest of the application can import it in a more centralized way. I understand that this is done when what is imported may change, and only one file needs to be changed. And that we will not stop using Stellar, obviously. If it's unnecessary, I'll remove it.

File: src/helpers/security.ts [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+};
+
+const keyPair = await generateKeyPair();
+
+export const exportCryptoKey = async (): Promise<ArrayBuffer> => {
+    const privateKey = keyPair.privateKey;
+
+    const keydata = await window.crypto.subtle.exportKey(
+        'pkcs8', //can be "jwk" (public or private), "spki" (public only), or "pkcs8" (private only)
+        privateKey!, //can be a publicKey or privateKey, as long as extractable was true
```
Comment (fsodano) [inline src/helpers/security.ts]: extra comments

File: src/helpers/security.ts [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+    return keyPair;
+};
+
+const keyPair = await generateKeyPair();
+
+export const exportCryptoKey = async (): Promise<ArrayBuffer> => {
+    const privateKey = keyPair.privateKey;
+
+    const keydata = await window.crypto.subtle.exportKey(
```
Comment (fsodano) [inline src/helpers/security.ts]: camelCase

File: src/helpers/security.ts [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        'pkcs8', //can be "jwk" (public or private), "spki" (public only), or "pkcs8" (private only)
+        privateKey!, //can be a publicKey or privateKey, as long as extractable was true
+    );
+
+    return keydata;
+};
+
+export const encryptPrivateKey = async (key: string) => {
```
Comment (fsodano) [inline src/helpers/security.ts]: missing return type

File: src/helpers/security.ts [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+            hash: 'SHA-256',
+        },
+        true,
+        ['encrypt', 'decrypt'],
+    );
+
+    return keyPair;
+};
+
+const keyPair = await generateKeyPair();
```
Comment (fsodano) [inline src/helpers/security.ts]: what's the purpose of  this?

**Review by fsodano — CHANGES_REQUESTED** [review]
The key management logic is spread out in 3 different files:
- connect
- storage
- security

storage should only know about storing things -- and it should hide the implementation details from other components. They should not care whether it's localStorage or sessionStorage.

security.ts should not care about storage, it only knows about encrypting and decrypting, and importing and exporting keys

File: src/routes/Connect.svelte [hunk 16]
```diff
@@ -12,8 +12,17 @@
...
-<button on:click="{initConnect}">Connect with private key</button>
-<button on:click="{decryptPk}">Decrypt PK</button>
+<h1>Public Key: {$publicKey ? $publicKey : 'Waiting connection...'}</h1>
+<button on:click="{() => ($isVisible = !$isVisible)}">Show key</button>
+
+{#if !$isVisible}
+    <input type="password" bind:value="{$inputValue}" />
+{:else}
+    <input type="text" bind:value="{$inputValue}" />
+{/if}
```
Comment (Ja-boop) [inline src/routes/Connect.svelte]: This is done because svelte doesn't want dynamic types when using bind:value

File: src/helpers/security.ts [hunk 17]
```diff
@@ -1,33 +1,31 @@
...
-    );
+import type { cryptoData } from '../interfaces/interface';
 
-    return keyPair;
+const uintEightArrayFromCharCode = (data: string): Uint8Array => {
+    const array = new Uint8Array([...data].map((char) => char.charCodeAt(0)));
+    return array;
 };
 
-const keyPair = await generateKeyPair();
```
Comment (Ja-boop) [inline src/helpers/security.ts]: Unnecessary call of a function

File: env.dist [hunk 18]
```diff
@@ -0,0 +1,2 @@
+VITE_TESTNET='https://horizon-testnet.stellar.org'
```
Comment (fsodano) [inline env.dist]: this should be .env.dist
also add a .env.prod.dist

Comment (fsodano) [inline env.dist]: VITE_HORIZON_URL

File: src/api/stellarSdk.ts (hunk 4 above)
Comment (fsodano) [reply src/api/stellarSdk.ts]: Please remove

File: src/helpers/security.ts [hunk 19]
```diff
@@ -0,0 +1,86 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
```
Comment (fsodano) [inline src/helpers/security.ts]: Configure this globally

File: src/helpers/security.ts [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+        hash: 'SHA-256',
+    },
+    true,
+    ['encrypt', 'decrypt'],
+);
+
+const exportCryptoKey = async (): Promise<string> => {
+    const cryptoKey = await window.crypto.subtle.exportKey('pkcs8', keyPair.privateKey!);
+    const cryptoKeyToUint8Array = new Uint8Array(cryptoKey);
+    const cryptoKeyString = String.fromCharCode.apply(null, Array.from(cryptoKeyToUint8Array)); // A type error will pop without Array.from
```
Comment (fsodano) [inline src/helpers/security.ts]: remove comment

File: src/helpers/storage.ts [hunk 21]
```diff
@@ -0,0 +1,14 @@
...
+export const storageData = async (key: string, cryptoKey: string) => {
+    sessionStorage.setItem('cryptoKey', cryptoKey);
+    sessionStorage.setItem('privateKey', key);
+};
+
+export const getStorageData = (): cryptoData => {
+    const secretKey = sessionStorage.getItem('privateKey');
+    const cryptoKey = sessionStorage.getItem('cryptoKey');
+
+    return { privateKey: secretKey!, cryptoKey: cryptoKey! };
```
Comment (fsodano) [inline src/helpers/storage.ts]: this should be an interface

File: src/helpers/storage.ts [hunk 22]
```diff
@@ -0,0 +1,14 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import type { cryptoData } from '../interfaces/interface';
```
Comment (fsodano) [inline src/helpers/storage.ts]: IStoredKey

File: src/interfaces/interface.ts [hunk 23]
```diff
@@ -0,0 +1,4 @@
+export interface cryptoData {
```
Comment (fsodano) [inline src/interfaces/interface.ts]: This file name should reflect the name of the interface, e.g. /IStoredKey.ts

File: src/helpers/storage.ts [hunk 24]
```diff
@@ -0,0 +1,14 @@
...
+
+export const storageData = async (key: string, cryptoKey: string) => {
+    sessionStorage.setItem('cryptoKey', cryptoKey);
+    sessionStorage.setItem('privateKey', key);
+};
+
+export const getStorageData = (): cryptoData => {
```
Comment (fsodano) [inline src/helpers/storage.ts]: prefer simple functions

File: src/helpers/storage.ts [hunk 25]
```diff
@@ -0,0 +1,14 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import type { cryptoData } from '../interfaces/interface';
+
+export const storageData = async (key: string, cryptoKey: string) => {
```
Comment (fsodano) [inline src/helpers/storage.ts]: export function storageData

Comment (fsodano) [reply src/helpers/storage.ts]: add return type (void)

File: src/routes/Connect.svelte [hunk 26]
```diff
@@ -1,4 +1,9 @@
-<script>
+<script lang="ts">
+    import { isVisible, inputValue } from '../stores/store';
+    import { connectWithStorageData, connectWithSecretKey } from '../helpers/connect';
```
Comment (fsodano) [inline src/routes/Connect.svelte]: remove the helper and use these functions here

File: src/helpers/connect.ts [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+};
+
+export const connectWithStorageData = async () => {
+    try {
+        const data = getStorageData();
+        const decryptedKey = await decryptPrivateKey(data.privateKey, data.cryptoKey);
+        await connectWithSecretKey(decryptedKey);
+    } catch (e) {
+        console.log('Private key was not found in Storage');
+        return e;
```
Comment (fsodano) [inline src/helpers/connect.ts]: this is a throw, not a return

File: src/helpers/connect.ts [hunk 28]
```diff
@@ -0,0 +1,29 @@
...
+    }
+};
+
+export const connectWithStorageData = async () => {
+    try {
+        const data = getStorageData();
+        const decryptedKey = await decryptPrivateKey(data.privateKey, data.cryptoKey);
+        await connectWithSecretKey(decryptedKey);
+    } catch (e) {
+        console.log('Private key was not found in Storage');
```
Comment (fsodano) [inline src/helpers/connect.ts]: `throw new StoredKeyNotFoundException`

File: src/helpers/connect.ts [hunk 29]
```diff
@@ -0,0 +1,29 @@
...
+
+export const connectWithSecretKey = async (key: string) => {
+    try {
+        const sourceKeys = StellarSDK.Keypair.fromSecret(key);
+        const encryptedData = await getEncryptedData(key);
+        storageData(encryptedData.privateKey, encryptedData.cryptoKey);
+        publicKey.set(sourceKeys.publicKey());
+    } catch (e) {
+        console.error(e);
+        publicKey.set('Invalid key, try again');
```
Comment (fsodano) [inline src/helpers/connect.ts]: Set an error in the store instead of doing this

File: src/routes/Connect.svelte [hunk 30]
```diff
@@ -1,4 +1,9 @@
-<script>
+<script lang="ts">
```
Comment (fsodano) [inline src/routes/Connect.svelte]: src/routes/Connect/Connect.svelte
src/routes/Connect/connectHelper.ts
src/routes/Connect/errors/StorageKeyNotFoundError.ts
src/routes/Connect/errors/InvalidPrivateKeyError.ts

File: src/routes/Connect.svelte [hunk 31]
```diff
@@ -1,4 +1,9 @@
...
+    import { isVisible, inputValue } from '../stores/store';
+    import { connectWithStorageData, connectWithSecretKey } from '../helpers/connect';
+    import { publicKey } from '../stores/store';
+
+    connectWithStorageData();
```
Comment (fsodano) [inline src/routes/Connect.svelte]:
```
try{
    const storedKey = getStoredKey(); // ex connectWithStorageData (only returns the IStoredKey or it throws an exception)
    setPrivateKey(storedKey.privateKey) // setPrivateKey is now contained inside Connect.svelte and it replaces connectWithSecretKey
}catch(e){
   // if e typeof StoredKeyNotFoundError (create custom exceptions)
}
```

File: src/routes/Connect.svelte [hunk 32]
```diff
@@ -7,3 +12,17 @@
     <li class="simple-signer xbull-wallet">Connect with xBull</li>
     <li class="simple-signer private-key-wallet">Connect with Private Key</li>
 </ul>
+<h1>Public Key: {$publicKey ? $publicKey : 'Waiting connection...'}</h1>
+<button on:click="{() => ($isVisible = !$isVisible)}">Show key</button>
```
Comment (fsodano) [inline src/routes/Connect.svelte]: try to avoid inline mutations

Comment (fsodano) [reply src/routes/Connect.svelte]: add docs to svelte reference where it says this is recommended

File: src/routes/Connect.svelte [hunk 33]
```diff
@@ -7,3 +12,17 @@
...
+
+{#if !$isVisible}
+    <input type="password" bind:value="{$inputValue}" />
+{:else}
+    <input type="text" bind:value="{$inputValue}" />
+{/if}
```
Comment (fsodano) [inline src/routes/Connect.svelte]: Always check positive condition first

File: src/stores/store.ts [hunk 34]
```diff
@@ -0,0 +1,5 @@
+import { writable } from 'svelte/store';
+
+export const isVisible = writable(false);
```
Comment (fsodano) [inline src/stores/store.ts]: isPublicKeyVisible

File: src/stores/store.ts [hunk 35]
```diff
@@ -0,0 +1,5 @@
...
+export const publicKey = writable('');
```
Comment (fsodano) [inline src/stores/store.ts]: This is ok to be public

**Review by fsodano — CHANGES_REQUESTED** [review]
Have a look at the comments provided, thank you!

File: src/routes/Connect.svelte (hunk 32 above)
Comment (Ja-boop) [reply src/routes/Connect.svelte]: https://svelte.dev/tutorial/inline-handlers

File: .env.dist [hunk 36]
```diff
@@ -0,0 +1,2 @@
+VITE_HORIZON_URL='https://horizon-testnet.stellar.org'
```
Comment (fsodano) [inline .env.dist]: no need for quotes

File: src/helpers/security.ts [hunk 37]
```diff
@@ -0,0 +1,85 @@
+import type { IStoredKeys } from '../routes/connect/IStoredKeys';
```
Comment (fsodano) [inline src/helpers/security.ts]: the interface name must be singular

File: src/helpers/security.ts [hunk 38]
```diff
@@ -0,0 +1,85 @@
...
+
+const uintEightArrayFromCharCode = (data: string): Uint8Array => {
```
Comment (fsodano) [inline src/helpers/security.ts]: uint8ArrayFromCharCode, and create this as a simple function instead of an arrow function, please

File: src/helpers/storage.ts [hunk 39]
```diff
@@ -0,0 +1,18 @@
+import type { IStoredKeys } from '../routes/connect/IStoredKeys';
+import StorageKeyNotFoundError from '../routes/connect/errors/StorageKeyNotFoundError';
+
+export function storageData(key: string, cryptoKey: string): void {
+    sessionStorage.setItem('cryptoKey', cryptoKey);
+    sessionStorage.setItem('privateKey', key);
+}
+
+export function getStorageData(): IStoredKeys {
```
Comment (fsodano) [inline src/helpers/storage.ts]: I can now see why you've named it in plural, but instead of calling it IStoredKeys call it IStoredKeyPair

File: src/routes/connect/Connect.svelte [hunk 40]
```diff
@@ -0,0 +1,41 @@
...
+
+    try {
+        const storageKeys = getStorageData();
+        (async () => {
+            const decryptedKey = await decryptPrivateKey(storageKeys.privateKey, storageKeys.cryptoKey);
+            setPrivateKey(decryptedKey);
+        })();
+    } catch (e) {
+        if (e instanceof StorageKeyNotFoundError) {
+            console.log('Key was not found in Storage');
```
Comment (fsodano) [inline src/routes/connect/Connect.svelte]: this should probably be shown to the user, but you can do it in another ticket.

File: src/routes/connect/connectStores.ts [hunk 41]
```diff
@@ -0,0 +1,5 @@
+import { writable } from 'svelte/store';
```
Comment (fsodano) [inline src/routes/connect/connectStores.ts]: why is connectStores a plural name? Isn't it just one store?

File: src/routes/connect/connectHelpers.ts [hunk 42]
```diff
@@ -0,0 +1,19 @@
+import { Keypair } from 'stellar-sdk';
+import { publicKey } from '../../stores/store';
+import { connectionError } from './connectStores';
```
Comment (fsodano) [inline src/routes/connect/connectHelpers.ts]: the helper should not access the store, it should only know how to get the private key or throw an error. The store should be accessed by the component.

**Review by fsodano — CHANGES_REQUESTED** [review]

File: src/helpers/storage.ts [hunk 43]
```diff
@@ -0,0 +1,18 @@
...
+    const secretKey = sessionStorage.getItem('privateKey');
+    const cryptoKey = sessionStorage.getItem('cryptoKey');
+
+    if (secretKey && cryptoKey) {
+        return { privateKey: secretKey, cryptoKey: cryptoKey };
+    } else {
+        throw new StorageKeyNotFoundError();
+    }
+}
```
Comment (fsodano) [inline src/helpers/storage.ts]: I just realised all of this doesn't belong in the storage helper. The storage helper shouldn't know what is it storing.

File: src/helpers/storage.ts [hunk 44]
```diff
@@ -0,0 +1,18 @@
+import type { IStoredKeys } from '../routes/connect/IStoredKeys';
+import StorageKeyNotFoundError from '../routes/connect/errors/StorageKeyNotFoundError';
+
+export function storageData(key: string, cryptoKey: string): void {
```
Comment (fsodano) [inline src/helpers/storage.ts]: `function store(key: string, value: any): void`

File: src/helpers/storage.ts (hunk 39 above)
Comment (fsodano) [reply src/helpers/storage.ts]: `function get(key: string): any`

File: src/routes/connect/Connect.svelte [hunk 45]
```diff
@@ -0,0 +1,41 @@
+<script lang="ts">
+    import { isPrivateKeyVisible, inputValue, connectionError } from './connectStores';
+    import { setPrivateKey } from './connectHelpers';
+    import { decryptPrivateKey } from '../../helpers/security';
+    import { publicKey } from '../../stores/store';
+    import { getStorageData } from '../../helpers/storage';
+    import StorageKeyNotFoundError from './errors/StorageKeyNotFoundError';
+
+    try {
+        const storageKeys = getStorageData();
```
Comment (fsodano) [inline src/routes/connect/Connect.svelte]: getStoredKeys is a function that should go on a global (src/helpers) keyManager.ts file

File: src/routes/connect/Connect.svelte [hunk 46]
```diff
@@ -0,0 +1,60 @@
...
+            const stellarKeyPair = await getStellarKeypair(key);
+            storeKey(key);
+            $publicKey = stellarKeyPair.publicKey();
+        } catch (e) {
+            if (e instanceof InvalidPrivateKeyError) {
+                $publicKey = '';
+                $setConnectionError = 'Invalid key, please try again';
+            }
+        }
+    }
```
Comment (fsodano) [inline src/routes/connect/Connect.svelte]: functions should be declared before they are used, you use this in the function above, so you should put this function first

File: src/routes/connect/connectStore.ts [hunk 47]
```diff
@@ -0,0 +1,5 @@
+import { writable } from 'svelte/store';
+
+export const isPrivateKeyVisible = writable(false);
+export const setInputValue = writable('');
+export const setConnectionError = writable('');
```
Comment (fsodano) [inline src/routes/connect/connectStore.ts]: these have function names, yet they are not functions. Simply call them inputValue and connectionError
